app/Models/GithubUser.php

```diff
@@ -3,75 +3,120 @@
+use Illuminate\Database\Eloquent\Relations\HasMany;
+
 /**
  * @property int|null $canonical_id
- * @property \Illuminate\Database\Eloquent\Collection $changelogEntries ChangelogEntry
+ * @property-read \Illuminate\Database\Eloquent\Collection<ChangelogEntry> $changelogEntries
  * @property \Carbon\Carbon|null $created_at
+ * @property-read string|null $created_at_json
  * @property int $id
  * @property \Carbon\Carbon|null $updated_at
- * @property User $user
+ * @property-read string|null $updated_at_json
+ * @property-read User|null $user
  * @property int|null $user_id
  * @property string|null $username
+ * @method static Builder withGithubInfo()
  */
 class GithubUser extends Model
 {
-    public static function importFromGithub($data)
+    /**
+     * Check if the app is capable of authenticating users via the GitHub API.
+     */
+    public static function canAuthenticate(): bool
+    {
+        return config('osu.github.client_id') !== null
+            && config('osu.github.client_secret') !== null;
+    }
+
+    /**
+     * Create or update a GitHub user with data from the GitHub API. Optionally
+     * associate the GitHub user to an osu! user.
+     */
+    public static function importFromGithub(array $apiUser, ?User $user = null): static
     {
-        $githubUser = static::where('canonical_id', '=', $data['id'])->first();
+        $params = [
+            'canonical_id' => $apiUser['id'],
+            'username' => $apiUser['login'],
+        ];
+        if ($user !== null) {
+            $params['user_id'] = $user->getKey();
+        }
+
+        $githubUser = static::where('canonical_id', $params['canonical_id'])->first()
+            ?? static::where('username', $params['username'])->last();
 
-        if (isset($githubUser)) {
-            $githubUser->update(['username' => $data['login']]);
+        if ($githubUser === null) {
+            return static::create($params);
         } else {
-            $githubUser = static::where('username', '=', $data['login'])->last();
-
-            if (isset($githubUser)) {
-                $githubUser->update(['canonical_id' => $data['id']]);
-            } else {
-                $githubUser = static::create([
-                    'canonical_id' => $data['id'],
-                    'username' => $data['login'],
-                ]);
-            }
+            $githubUser->update($params);
+            return $githubUser;
         }
+    }
 
-        return $githubUser;
+    public function changelogEntries(): HasMany
+    {
+        return $this->hasMany(ChangelogEntry::class);
     }
 
-    public function user()
+    public function user(): BelongsTo
     {
         return $this->belongsTo(User::class, 'user_id');
     }
 
-    public function changelogEntries()
+    public function scopeWithGithubInfo(Builder $query): void
     {
-        return $this->hasMany(ChangelogEntry::class);
+        $query->whereNotNull(['canonical_id', 'username']);
```

Comment:
looking at actual data, there's no entry with either set to null

Reply:
are there any with a blank username? there's some function that looks like it was trying to handle blank or null username

Reply:
nope